### PR TITLE
idris: fix modules

### PR DESCRIPTION
**Pull request by brainrake**
###### Motivation for this change

Some Idris modules (eg. `lightyear`, the non-builtin ones) were broken because they tried to use the `meta.platforms` attribute from the `idris` they were called with, but that was changed to a wrapper that didn't include the meta attribute, so the modules would not build with a missing attribute error. 
This PR reuses Idris' `meta.platforms` variable in the wrapper, so the modules build successfully.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by joachifm** [comment]
cc @mpickering

**Comment by mpickering** [comment]
@brainrape Perhaps you can try the heavily refactored infrastructure I have submitted in #31327 ?

I think this problem is fixed there.

**Comment by mpickering** [comment]
@brainrape It worked properly last time I tried it? In what sense do you mean it is stalled?

**Comment by brainrape** [comment]
@mpickering I support #31327, appreciate the work done there, and would like to see it merged. However, I got the impression that it is not making progress towards being merged (changes requested, possibly further review required, conflict with master). In the mean time we could have working Idris modules in `master` with this small, easy to review PR.

**Comment by mpickering** [comment]
OK, I can't merge this but #31327 is finished from my perspective. I'll try to rebase it now.

**Comment by brainrape** [comment]
Thank you! Here's hoping it will make it all the way in this time. I'll close this when it does.

**Comment by mpickering** [comment]
Your change was already merged.

https://github.com/NixOS/nixpkgs/commit/69d4adee1f8b6278a19708a9051f3290f8f13e04

**Comment by brainrape** [comment]
@mpickering I'm not sure what you mean, but eg. lightyear and other idris libraries are broken in current master and need this PR to be fixed. Build output below:
```
└──> git show-ref HEAD
cff5eec3825f6158f8296eb194c2b1d53d36cad8 refs/remotes/origin/HEAD
└──> nix-shell -f . --attr idrisPackages.lightyear
error: attribute ‘platforms’ missing, at /home/ssdd/dev/nixpkgs/pkgs/development/idris-modules/lightyear.nix:25:11
(use ‘--show-trace’ to show detailed location information)
```

**Comment by mpickering** [comment]
Yes sorry, my comment was wrong.

**Comment by mpickering** [comment]
I rebased my other branch now if you want to try that for your use case.

**Comment by brainrape** [comment]
I checked out your branch and the Idris modules are still broken, and fixed by this PR in the same way. In conclusion, the two PRs are separate and this should definitely be merged in any case.
@shlevy could you do the honors? Thanks!

**Comment by mpickering** [comment]
How are you building the packages? Are you using `with-packages`?

**Comment by brainrape** [comment]
I included the build command and output, two comments up.

**Comment by brainrape** [comment]
Yes, eventually, I would use that derivation as an argument to `with-packages` in my project. But evaluating the derivation fails before that.

**Comment by mpickering** [comment]
Your command doesn't work for me.

I then tried `nix-shell -p idrisPackages.lightyear` which works fine for me locally. 

87636ac3115920c71415f47ba6390031b01884d8git

**Comment by brainrape** [comment]
To actually try to build the lightyear package (and not just start a shell in the build env as above), I'd use `nix-build /path/to/nixpkgs --attr idrisPackages.lightyear`

**Comment by mpickering** [comment]
yes, that works fine for me. 

`nix-shell -p` does build `idrisPackage.lightyear`, it creates a shell with `idrisPackage.lightyear` available. 

You can see the platforms attribute is set in the lightyear deviation?

**Comment by brainrape** [comment]
You have to add `-f /path/to/nixpkgs`, otherwise `nix-shell` will use `NIX_PATH` which is usually managed by `nix-channel`, possibly `nixos-17.10` when it still worked, instead of your local checkout of `nixpkgs`.

**Comment by mpickering** [comment]
My `NIX_PATH` points to my local `nixpkgs`.

```
[root@nixos:~/devel/nixpkgs]# echo $NIX_PATH
nixpkgs=/root/devel/nixpkgs:nixos-config=/etc/nixos/configuration.nix
```

**Comment by brainrape** [comment]
@mpickering yes, `lightyear.nix` sets `platforms`, but by inheriting from `idris.meta`, which _doesn't_ have a `platforms` attribute, which is exactly what this PR fixes!

**Comment by brainrape** [comment]
Not sure why you can't reproduce.

**Comment by mpickering** [comment]
Which platform are you on?

**Comment by mpickering** [comment]
I can't find anther use of `symlinkJoin` which sets any `meta` attribute.

**Comment by brainrape** [comment]
My platform is NixOS, as stated in the PR description. x86_64. I don't get it.. that attribute is really missing, and any package that uses `idris-wrapper.nix` and tries to inherit its `platforms` will surely fail, plain as day.
Not sure how `symlinkJoin` fits in the picture... I read the source at https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix#L76 and it just passes on the args you give it, more or less.

Here's the story again, if it wasn't clear from the description. There's an `idris` package that has a `platforms` attribute. Lightyear inherited the platforms attribute from it. But then `idris-wrapper` was created to address  #10450 and replaced `idris` in lightyear etc. It did not define a `platforms` attribute. So the the packages that tried to inherit it, like lightyear, failed with attribute not found. So I added the attribute.

I suspect you're somehow not building the right nixpkgs. I don't see how anything else (eg. sandbox) could affect this, I traced the evaluation and couldn't find any way to divine `platforms`.

This discussion is getting a bit out of hand for a straightforward one-line change, and I'm sorry to mention this but a lot of the above comments didn't check out. I'm convinced though that we're both acting in good faith and want the same thing (working nixpkgs) so maybe we need a fresh perspective. Let's let someone else take a look. @joachifm could you check this out please?

Basically:
```
nix-build /path/to/nixpkgs --attr idrisPackages.lightyear
```
It should fail with `error: attribute ‘platforms’ missing`, and this PR will fix it so should be merged. If not, I missed something weird, but it still fails for me, and we should find out why.

**Comment by mpickering** [comment]
@Infinisil tried it on IRC and got the same error message as you. I really wish I could reproduce this.. I have tried on master and my branch.

**Comment by mpickering** [comment]
@GrahamcOfBorg build idrisPackages.lightyear

**Comment by Infinisil** [comment]
I am worried we're having some problems with non-reproducibility, reported by @FRidh earlier today: https://github.com/NixOS/nixpkgs/issues/33362#issuecomment-362820639

**Comment by 7c6f434c** [comment]
@GrahamcOfBorg build idrisPackages.lightyear

**Comment by brainrape** [comment]
Ok that's weird. Maybe try breaking it and see what happens. Change smth and see if it has an effect. Add debug output. (`lib.trace` or similar iirc)
Anyway, it's a minority report and you have a telepathic nix :)
So this PR is good to merge.

**Comment by brainrape** [comment]
^ is this on this branch or master?

**Comment by 7c6f434c** [comment]
On this branch

**Comment by mpickering** [comment]
@brainrape on this branch.

I think I can't reproduce this because I have "allowBroken" in my config but I can't work out how to properly disable it.

**Comment by 7c6f434c** [comment]
Interesting… I don't know enough about Idris to understand why the error from Darwin failure doesn't occur on Linux.

**Comment by mpickering** [comment]
@brainrape Thank you very much for your patience. I can now reproduce this after unsetting `allowBroken`. 

@7c6f434c I think the failure is not to do with this patch but an already existing issue.

**Comment by 7c6f434c** [comment]
Yes, but you may want to set a more narrow `meta.platforms`

**Comment by mpickering** [comment]
The right place would be on the haskell packages which are failing to build. I can't work out why they are though. Merging this is the right thing to do for now I think, however in future I think it would be better to set meta.platforms on a per idris module level as idris modules can call C FFI like haskell modules.

**Comment by brainrape** [comment]
Thank you for all your help in elucidating this issue.
